File: api/models/data/activity_log.py
from django.db import models
from django.contrib.contenttypes.models import ContentType
from django.contrib.contenttypes.fields import GenericForeignKey
from account.models import User
from api.models.data.base import BaseModel



class ActivityLog(BaseModel):
    ACTION_CHOICES = [
        ('create', 'Create'),
        ('update', 'Update'),
        ('delete', 'Delete'),
        ('login', 'Login'),
        ('logout', 'Logout'),
        ('view', 'View'),
        ('export', 'Export'),
        ('payment', 'Payment'),
    ]
    
    user = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, related_name='activity_logs')

    action = models.CharField(max_length=20, choices=ACTION_CHOICES)
    model_name = models.CharField(max_length=100)
    object_id = models.PositiveIntegerField(null=True, blank=True)
    content_type = models.ForeignKey(ContentType, on_delete=models.SET_NULL, null=True, blank=True)
    content_object = GenericForeignKey('content_type', 'object_id')
    description = models.TextField()
    ip_address = models.GenericIPAddressField(null=True, blank=True)
    user_agent = models.TextField(null=True, blank=True)
    changes = models.JSONField(null=True, blank=True)
    
    class Meta:
        ordering = ['-created_at']
        indexes = [
            models.Index(fields=['-created_at']),
            models.Index(fields=['user', '-created_at']),

            models.Index(fields=['action', '-created_at']),
        ]
    
    def __str__(self):
        user_name = f"{self.user.first_name} {self.user.last_name}".strip() if self.user else "Unknown"
        return f"{user_name} - {self.action} - {self.model_name} at {self.created_at}"

File: frontend/src/types/activity-log.ts
export interface ActivityLog {
  id: number;
  user: number;
  user_name: string;
  user_email: string;
  user_role: string;
  company: number | null;
  company_name: string | null;
  action: 'create' | 'update' | 'delete' | 'login' | 'logout' | 'view' | 'export' | 'payment';
  model_name: string;
  object_id: number | null;
  description: string;
  ip_address: string | null;
  user_agent: string | null;
  changes: Record<string, any> | null;
  created_at: string;
  updated_at: string;
}

export interface ActivityLogStats {
  total_activities: number;
  by_action: Array<{ action: string; count: number }>;
  by_user: Array<{ 
    user__username: string; 
    user__first_name: string; 
    user__last_name: string; 
    user__role: string;
    count: number 
  }>;
  by_model: Array<{ model_name: string; count: number }>;
}
